--- a/src/avx2/deser.rs
+++ b/src/avx2/deser.rs
@@ -179,20 +179,11 @@ impl<'de> Deserializer<'de> {
                     src_i += s;
                     dst_i += o;
                 } else {
-                    // simple 1:1 conversion. Will eat bs_dist+2 characters in input and
-                    // write bs_dist+1 characters to output
-                    // note this may reach beyond the part of the buffer we've actually
-                    // seen. I think this is ok
-                    let escape_result: u8 =
-                        unsafe { *ESCAPE_MAP.get_kinda_unchecked(escape_char as usize) };
-                    if escape_result == 0 {
-                        return Err(Self::error_c(src_i, escape_char as char, InvalidEscape));
-                    }
-                    unsafe {
-                        *buffer.get_kinda_unchecked_mut(dst_i + bs_dist as usize) = escape_result;
-                    }
                     src_i += bs_dist as usize + 2;
-                    dst_i += bs_dist as usize + 1;
+                    dst_i += bs_dist as usize + 2;
+                    unsafe {
+                        *buffer.get_kinda_unchecked_mut(dst_i-1) = *src.get_kinda_unchecked(src_i-1);
+                    };
                 }
             } else {
                 // they are the same. Since they can't co-occur, it means we encountered
